tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+ * @author
+ *   Jure Kukovec
+ */
+class ItfToTla[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T]) {
+
+  /**
+   * Checks that the Apalache-specific ITF requirements are met, as defined in ADR15. If yes, returns a mapping from
+   * variables to their types.
+   */
+  private[itf] def validateShapeAndGetTypes(json: T): Map[String, TlaType1] = {
+    val varTypesOpt = for {
+      meta <- json.getFieldOpt(META_FIELD)
+      varTypes <- meta.getFieldOpt(VAR_TYPES_FIELD)
+    } yield varTypes
+    val varTypesJson = varTypesOpt.getOrElse {
+      throw new JsonDeserializationError("ITF JSON does not satisfy Apalache requirements. " +
+        "Toplevel \"#meta\" field must exist and contain \"varTypes\". " +
+        "See ADR15: https://apalache.informal.systems/docs/adr/015adr-trace.html#itf-as-input-to-apalache. ")
+    }
+
+    val diffOpt = for {
+      vars <- json.getFieldOpt(VARS_FIELD)
+      varTypesKeys <- varTypesJson.allFieldsOpt
+    } yield {
+      val varsSet = scalaFactory.asSeq(vars).map { scalaFactory.asStr }.toSet
+      // Symmetric difference
+      (varsSet -- varTypesKeys) ++ (varTypesKeys -- varsSet)
+    }
+
+    val diff = diffOpt.getOrElse {
+      throw new JsonDeserializationError(s"\"$VARS_FIELD\" or \"$VAR_TYPES_FIELD\" contents are malformed.")
+    }
+    if (diff.nonEmpty)
+      throw new JsonDeserializationError(s"\"$VARS_FIELD\" and \"$VAR_TYPES_FIELD\" define different states. " +
+        s"Variable(s) ${diff.toSeq.mkString(", ")} present in exactly one of the two fields.")
+
+    varTypesJson.allFieldsOpt.get.map { k =>
+      k -> DefaultType1Parser.parseType(scalaFactory.asStr(varTypesJson.getFieldOpt(k).get))
+    }.toMap
+  }
+
+  /**
+   * Some values, like the predefined sets (Int, Nat, etc.) are encoded as "UNSERIALIZABLE_FIELD": `<value>` Currently,
+   * per our discussions, we disallow the use of this tag, but the function can be extended in the future.
+   *
+   * E.g. we could return Some(tla.intSet()) on str = "Int".
+   */
+  private[itf] def attemptUnserializable(json: T): Option[TBuilderInstruction] =
+    if (json.allFieldsOpt.contains(Set(UNSERIALIZABLE_FIELD))) {
+      val str = scalaFactory.asStr(json.getFieldOpt(UNSERIALIZABLE_FIELD).get)
+      throw new JsonDeserializationError(s"Unserializable value $str is disallowed.")
+    } else None
+
+  // Note: attemptUnserializable + getOrElse is future-proof, and allows us to eventually case-match
+  // unserializable values.
+
+  private[itf] def typeDrivenBuild(json: T, tt1: TlaType1): TBuilderInstruction =
+    attemptUnserializable(json).getOrElse {
+      tt1 match {
+        case BoolT1 =>
+          tla.bool(scalaFactory.asBool(json))
+        case StrT1 =>
+          val s = scalaFactory.asStr(json)
+          if (ModelValueHandler.isModelValue(s))
+            throw new JsonDeserializationError(
+                s"$s is annotated as a string, but has the value of an uninterpreted literal.")
```

Comment:
This is definitely not a JSON deserialization error. This is a type checking error :D

Reply:
Yes, in the provided json, which is why the json can't be read into a trace.